README
NAME
    Dist::Zilla::Plugin::CheckExtraTests - check xt tests before release

VERSION
    version 0.025

SYNOPSIS
    In your dist.ini:

      [CheckExtraTests]

DESCRIPTION
    Runs all xt tests before release. Dies if any fail. Sets RELEASE_TESTING
    and AUTHOR_TESTING.

    If you use Dist::Zilla::Plugin::TestRelease, you should consider using
    Dist::Zilla::Plugin::RunExtraTests instead, which enables xt tests to
    run as part of "[TestRelease]" and is thus a bit more efficient as the
    distribution is only built once for testing.

ATTRIBUTES
  default_jobs
    This attribute is the default value that should be used as the "jobs"
    argument for prerelease tests.

SEE ALSO
    *   Dist::Zilla

SUPPORT
  Bugs / Feature Requests
    Please report any bugs or feature requests through the issue tracker at
    <https://github.com/dagolden/Dist-Zilla-Plugin-CheckExtraTests/issues>.
    You will be notified automatically of any progress on your issue.

  Source Code
    This is open source software. The code repository is available for
    public review and contribution under the terms of the license.

    <https://github.com/dagolden/Dist-Zilla-Plugin-CheckExtraTests>

      git clone https://github.com/dagolden/Dist-Zilla-Plugin-CheckExtraTests.git

AUTHORS
    *   David Golden <dagolden@cpan.org>

    *   Jesse Luehrs <doy@cpan.org>

CONTRIBUTORS
    *   Christopher J. Madsen <cjm@cpan.org>

    *   Karen Etheridge <ether@cpan.org>

    *   Kent Fredric <kentfredric@gmail.com>

    *   Olivier Mengue <dolmen@cpan.org>

    *   Ricardo Signes <rjbs@cpan.org>

COPYRIGHT AND LICENSE
    This software is Copyright (c) 2014 by David Golden.

    This is free software, licensed under:

      The Apache License, Version 2.0, January 2004



Changes
Revision history for Dist-Zilla-Plugin-CheckExtraTests

0.025     2014-11-21 11:29:50-05:00 America/New_York

    [Changed]

    - Respects dzil verbose flag

0.024     2014-11-04 05:30:13-05:00 America/New_York

    [Added]

    - RunExtraTests now also runs miscellaneous xt/ tests

0.023     2014-10-30 16:25:03-04:00 America/New_York

    [Changed]

    - Removed unused Moose::Autobox

0.022     2014-07-14 17:36:05-04:00 America/New_York

    - [RunExtraTests] now accepts the --jobs flag passed to 'dzil test' and
      'dzil release'

0.021     2014-03-25 14:34:02-04:00 America/New_York

    [Fixed]

    - specifies Path::Tiny prerequisite of 0.013 or better

0.020     2014-03-22 23:12:58-04:00 America/New_York

    [Fixed]

    - default jobs support was buggy in the last version

0.019     2014-03-22 22:48:04-04:00 America/New_York

    [Changed]

    - Added 'jobs' argument for 'dzil xtest' for parallel testing

    - passing new 'default_jobs' option (added to Dist::Zilla 5.014) through
      to test runner for parallel testing

0.018     2014-02-18 17:20:58-05:00 America/New_York

    [Fixed]

    - CheckExtraTest now force removes temp directory, the way
      Path::Class used to do.

0.017     2014-02-13 15:03:55-05:00 America/New_York

    [Changed]

    - Removed dependency on Path::Class in favor of Path::Tiny

0.016     2013-11-15 11:33:06 America/New_York

    [Changed]

    - the 'xtest' command now accepts the same options as 'test' does

0.015     2013-10-31 15:03:13 America/New_York

    - Shipping as a stable release

0.014     2013-10-23 19:52:43 America/New_York

    [Fixed]

    - [CheckExtraTest] tests in a .build directory instead of the build
      directory

    - the 'xtest' command now uses Dist::Zilla methods to build a temporary
      directory, which also sets the .build/latest symlink correctly

0.013     2013-10-17 17:26:26 America/New_York

    [Changed]

    - [RunExtraTests] now runs xt/author ahead of xt/smoke and xt/release
      (Karen Etheridge)

0.012     2013-10-15 23:15:02 America/New_York

    [Changed]

    - 'dzil xtest' arguments now match as regex patterns for more
      flexibility

    [Fixed]

    - use all builders to build, not just the first (Karen Etheridge)

0.011     2013-02-09 20:53:59 America/New_York

    - Doc fixes (ETHER)

    - Path::Tiny fix (ETHER)

0.010     2013-02-07 23:09:15 America/New_York

    - Bump Dist::Zilla prereq to something modernish

0.009     2013-02-06 16:15:30 America/New_York

    - Swap Path::Class for Path::Tiny and Path::Class::Rule
      for Path::Iterator::Rule

0.008     2013-01-25 15:23:15 America/New_York

    - lazy loading of some dependencies (DOLMEN)

0.007     2012-05-31 12:00:29 America/New_York

    - change RunExtraTests plugin to avoid rebuilding the dist when
      it's not necessary (e.g. MakeMaker already ran the normal tests)
      (thanks, Christopher J. Madsen)

0.006     2012-05-10 23:08:11 America/New_York

    - run tests on the built dist, so that things like XS will work

    - add RunExtraTests plugin, which runs the appropriate xt tests when "dzil
      test" is run (but only with the appropriate flags).

0.005     2012-05-03 17:30:08 America/New_York

    - add 'xt' as command alias for 'xtest'

    - 'dzil xtest' now takes optional arguments that are
      treated as xt test filenames to run.  These will
      be found however deeply they are nested in xt.

0.004     2010-04-08 14:51:28 EST5EDT

    - manually set prereqs instead of using
      auto-detection to avoid a Dist::Zilla
      BUILD_REQUIRES bug

0.003     2010-04-07 14:02:34 EST5EDT

    - updated tests for Dist::Zilla 2.000+

    - switched from File::chdir to File::pushd

0.002     2010-03-23 19:28:48 EST5EDT

    - fixed problem in release check

0.001     2010-03-22 01:10:12 EST5EDT


    - first draft


t/lib/Test/DZil.pm
# 
# This file is part of Dist-Zilla-Plugin-CheckExtraTests
# 
# This software is Copyright (c) 2010 by David Golden.
# 
# This is free software, licensed under:
# 
#   The Apache License, Version 2.0, January 2004
# 
use strict;
use warnings;
package Test::DZil;

use Dist::Zilla::Tester;
use Params::Util qw(_HASH0);

use Sub::Exporter -setup => {
  exports => [
    dist_ini   => \'_dist_ini',
    simple_ini => \'_simple_ini',
  ],
  groups  => [ default => [ qw(dist_ini simple_ini) ] ],
};

sub _build_ini_builder {
  my ($starting_core) = @_;
  $starting_core ||= {};

  sub {
    my (@arg) = @_;
    my $new_core = _HASH0($arg[0]) ? shift(@arg) : {};

    my $core_config = { %$starting_core, %$new_core };

    my $config = '';

    for my $key (keys %$core_config) {
      my @values = ref $core_config->{ $key }
                 ? @{ $core_config->{ $key } }
                 : $core_config->{ $key };

      $config .= "$key = $_\n" for grep {defined} @values;
    }

    $config .= "\n" if length $config;

    for my $line (@arg) {
      my @plugin = ref $line ? @$line : ($line, {});
      my $moniker = shift @plugin;
      my $name    = _HASH0($plugin[0]) ? undef : shift @plugin;
      my $payload = shift(@plugin) || {};

      die "TOO MANY ARGS TO PLUGIN GAHLGHALAGH" if @plugin;

      $config .= '[' . $moniker;
      $config .= ' / ' . $name if defined $name;
      $config .= "]\n";

      for my $key (keys %$payload) {
        my @values = ref $payload->{ $key }
                   ? @{ $payload->{ $key } }
                   : $payload->{ $key };

        $config .= "$key = $_\n" for @values;
      }

      $config .= "\n";
    }

    return $config;
  }
}

sub _dist_ini {
  _build_ini_builder;
}

sub _simple_ini {
  _build_ini_builder({
    name     => 'DZT-Sample',
    abstract => 'Sample DZ Dist',
    version  => '0.001',
    author   => 'E. Xavier Ample <example@example.org>',
    license  => 'Perl_5',
    copyright_holder => 'E. Xavier Ample',
  });
}

1;


t/check-pass/lib/Foo.pm
# 
# This file is part of Dist-Zilla-Plugin-CheckExtraTests
# 
# This software is Copyright (c) 2010 by David Golden.
# 
# This is free software, licensed under:
# 
#   The Apache License, Version 2.0, January 2004
# 
package Foo;
our $VERSION = '0.002';
use strict;
use warnings;
# ABSTRACT: Foo this thing

1;

=head1 DESCRIPTION

Foo the foo.

=cut


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.019.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker 6.17;



my %WriteMakefileArgs = (
  "ABSTRACT" => "check xt tests before release",
  "AUTHOR" => "David Golden <dagolden\@cpan.org>, Jesse Luehrs <doy\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.17"
  },
  "DISTNAME" => "Dist-Zilla-Plugin-CheckExtraTests",
  "EXE_FILES" => [],
  "LICENSE" => "apache",
  "NAME" => "Dist::Zilla::Plugin::CheckExtraTests",
  "PREREQ_PM" => {
    "App::Prove" => "3.00",
    "Archive::Tar" => 0,
    "Dist::Zilla" => "4.3",
    "Dist::Zilla::App" => 0,
    "Dist::Zilla::Role::BeforeRelease" => 0,
    "Dist::Zilla::Role::TestRunner" => 0,
    "File::pushd" => 0,
    "Moose" => 2,
    "Moose::Autobox" => 0,
    "Path::Iterator::Rule" => 0,
    "Path::Tiny" => "0.013",
    "namespace::autoclean" => "0.09",
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Capture::Tiny" => 0,
    "Dist::Zilla::App::Tester" => 0,
    "Dist::Zilla::Tester" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec::Functions" => 0,
    "List::Util" => 0,
    "Params::Util" => 0,
    "Sub::Exporter" => 0,
    "Test::More" => "0.88",
    "Test::Requires" => 0,
    "Try::Tiny" => 0,
    "lib" => 0,
    "version" => 0
  },
  "VERSION" => "0.022",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "App::Prove" => "3.00",
  "Archive::Tar" => 0,
  "Capture::Tiny" => 0,
  "Dist::Zilla" => "4.3",
  "Dist::Zilla::App" => 0,
  "Dist::Zilla::App::Tester" => 0,
  "Dist::Zilla::Role::BeforeRelease" => 0,
  "Dist::Zilla::Role::TestRunner" => 0,
  "Dist::Zilla::Tester" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec::Functions" => 0,
  "File::pushd" => 0,
  "List::Util" => 0,
  "Moose" => 2,
  "Moose::Autobox" => 0,
  "Params::Util" => 0,
  "Path::Iterator::Rule" => 0,
  "Path::Tiny" => "0.013",
  "Sub::Exporter" => 0,
  "Test::More" => "0.88",
  "Test::Requires" => 0,
  "Try::Tiny" => 0,
  "lib" => 0,
  "namespace::autoclean" => "0.09",
  "strict" => 0,
  "version" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



